aiohttp/web_urldispatcher.py

```diff
@@ -979,8 +986,23 @@ def __init__(self) -> None:
         super().__init__()
         self._resources: List[AbstractResource] = []
         self._named_resources: Dict[str, AbstractResource] = {}
+        self._resource_index: dict[str, list[AbstractResource]] = {}
 
     async def resolve(self, request: Request) -> UrlMappingMatchInfo:
+        url_parts = request.rel_url.raw_parts
+        resource_index = self._resource_index
+        # Walk the url parts looking for candidates
+        for i in range(len(url_parts), 0, -1):
+            url_part = "/" + "/".join(url_parts[1:i])
+            if (resource_candidates := resource_index.get(url_part)) is not None:
+                for candidate in resource_candidates:
+                    if (
+                        match_dict := (await candidate.resolve(request))[0]
+                    ) is not None:
+                        return match_dict
+
+        # We didn't find any candidates, so we fallback to a linear search
```

Comment:
Is there a situation where this is going to find a result if we didn't already find one?

I think if this is kept it's going to make performance worse, as any 404s are going to take even longer (and there's plenty of bots looking for non-existent pages).

Reply:
> Is there a situation where this is going to find a result if we didn't already find one?
> 
> I think if this is kept it's going to make performance worse, as any 404s are going to take even longer (and there's plenty of bots looking for non-existent pages).

It should only have to hit this for the `add_domain` case. The performance will just a tiny bit be worse on a miss because it will have to do `O(n) + O(url_parts)` instead of `O(n)`.

Reply:
I adjusted it so only `MatchedSubAppResource` resources have to pay the price of the linear search

Reply:
3d0b0f952bd5c2b144de7f6043051782c48e7ddf

Reply:
I will add before and after bench marks for miss as well

Reply:
miss is now fast as well

```
resolve_time_first_url: 0.002494540996849537
resolve_time_last_url: 0.0025858330773189664
resolve_time_miss: 0.007047250051982701
```

Reply:
still need to fix system resources (working on that now)